persist/storer.go

```
@@ -3,168 +3,222 @@ package persist
+	return baseDir
 }
 
-// NewInMemoryStore will create and initialize an in-memory Storer.
+// NewInMemoryStore will create and initialize an in-memory Storer (not persistent).
 func NewInMemoryStore() Storer {
 	return &inMemoryStore{
-		Data:       make(map[string]interface{}),
+		cachedData: make(map[string]jsonEntry),
+		Data:       make(map[string][]byte),
 		Timestamps: make(map[string]int64),
 	}
 }
 
-// NewFileStore will create and initialize a disk-backed Storer.
-func NewFileStore(storePath string, l log.Logger) (Storer, error) {
-	store := &fileStore{
-		inMemoryStore: *NewInMemoryStore().(*inMemoryStore),
-		path:          storePath,
+// NewFileStore returns a disk-backed Storer using the provided file path
+func NewFileStore(storagePath string, ilog log.Logger, ttl time.Duration) (Storer, error) {
+	ms := NewInMemoryStore().(*inMemoryStore)
+
+	store := fileStore{
+		path:          storagePath,
+		ilog:          ilog,
+		inMemoryStore: *ms,
 	}
 
-	// Create the external directory for user-generated json
-	storeDir := filepath.Dir(store.path)
-	if _, err := os.Stat(storeDir); err != nil {
-		if err = os.MkdirAll(storeDir, dirFilePerm); err != nil {
-			return nil, fmt.Errorf("store directory in %s could not be created", storeDir)
+	if stat, err := os.Stat(storagePath); err == nil {
+		if now().Sub(stat.ModTime()) > ttl {
+			ilog.Debugf("store file (%s) is older than %v, skipping loading from disk.", storagePath, ttl)
+			return &store, nil
 		}
-	}
 
-	stat, err := os.Stat(store.path)
-	// Store file doesn't exist yet
-	if err != nil {
-		if _, err = os.OpenFile(store.path, os.O_CREATE|os.O_WRONLY, filePerm); err != nil {
-			return nil, fmt.Errorf("store directory not writable: %s", storeDir)
+		store.loadFromDisk()
+	} else if os.IsNotExist(err) {
+		folder := path.Dir(storagePath)
+		err := os.MkdirAll(folder, dirFilePerm)
+		if err != nil {
+			ilog.Errorf("can't create storage directory %q: %s", folder, err)
```

This should return an error (also if returned then it should not be logged at this level)